### PR TITLE
Zadanie2 na polskich firmach

### PR DESCRIPTION
**Pull request by Rafzzz3**
[PolishCompanies.ipynb](https://github.com/user-attachments/files/31156717/PolishCompanies.ipynb)

### CODE REVIEW

**Review by KLTPL — COMMENTED** [review]
Wydaje się że bardzo dobrze wykonane. Ciekawie zobaczyć inny dataset. Zabrakło mi mocno analizy zbioru, takiego wprowadzenia na poziomie EDA. Poza tym ładne wykresy, część dotycząca trenowania fajnie poprowadzona.

**Review by wgrzyb271 — APPROVED** [review]
Porządne wykonanie, nie mam się do czego przyczepić. Ciekawym rozwiązaniem było by też hybrydowe wykorzystanie optuny i gridsearch, ale to takie moje małe wtrącenie hehe. Co do wyboru zbioru danych, mam nadzieję że nie przewidzi mojego bankructwa 🫣

**Review by Miloproplayer — COMMENTED** [review]
Wydaje mi się, że fachowo wykonane zadanie. Dobry dobór metryk do problemu.

**Comment by PilarToZiomal** [comment]
Szczerze nie mam chyba do czego się przyczepić do kodu, więc przyczepie się merytorycznie XD:

1. **Cross-validation nie oznacza automatycznie braku overfittingu**

   W podsumowaniu piszesz, że dzięki użyciu walidacji krzyżowej modele się nie przeuczyły. Cross-validation nie gwarantuje braku overfittingu - pozwala lepiej oszacować zdolność modelu do generalizacji i pomaga dobrać hiperparametry, ale nadal możliwe jest przeuczenie modelu.

3. **Wnioski dotyczące XGBoost są zbyt optymistyczne**

   XGBoost uzyskał wysokie `precision`:

   ```text
   Precision: 0.800
   ```

   ale jednocześnie stosunkowo niskie:

   ```text
   Recall: 0.485
   ```

   Oznacza to, że jeśli model przewidzi bankructwo, często ma rację, ale wykrywa mniej niż połowę rzeczywistych przypadków bankructwa (dzięki Chat że umiesz to opisać XD).

   Stwierdzenie, że model „naprawdę dobrze się sprawdził”, jest trochę zbyt mocne. Warto w podsumowaniu zaznaczyć kompromis między wysokim precision a niskim recall.

4. **Random Forest i XGBoost nie są tym samym typem „lasu”**

   Random Forest i XGBoost wykorzystują drzewa decyzyjne, ale działają w inny sposób.

   * `RandomForestClassifier` wykorzystuje bagging - wiele drzew jest trenowanych niezależnie, a ich wyniki są następnie agregowane.
   * XGBoost wykorzystuje gradient boosting - drzewa powstają sekwencyjnie, a każde kolejne próbuje poprawić błędy poprzednich.

5. **Strojenie hiperparametrów nie jest w pełni reprodukowalne**

   W finalnych modelach używasz `random_state`, ale podczas strojenia z Optuną modele tworzysz wewnątrz funkcji.

   ```python
   model = XGBClassifier(**params)
   ```

   oraz analogicznie dla Random Foresta.

   Dodatkowo sampler Optuny również nie ma ustawionego seeda. W efekcie ponowne uruchomienie notebooka może dać inne wyniki strojenia i inne najlepsze hiperparametry.

   Warto ustawić `random_state` w modelach oraz seed dla samplera Optuny, np.:

   ```python
   sampler = optuna.samplers.TPESampler(seed=0)
   study = optuna.create_study(
       direction='maximize',
       sampler=sampler
   )
   ```

ZALICZONE na piątkę
